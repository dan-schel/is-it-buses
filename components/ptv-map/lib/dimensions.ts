export const lineWidth = 4;

export const interchangeThinLineWidth = 1;
export const interchangeThickLineWidth = 4;
export const interchangeBorderWidth = 1;
export const interchangeEdgeOffset = lineWidth / 2;
export const interchangeInnerOffset = 1;

export const terminusLineWidth = 3;
export const terminusExtents = 5;
