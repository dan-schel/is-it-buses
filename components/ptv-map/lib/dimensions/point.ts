export type InformalPoint = Point | { x: number; y: number };

export class Point {
  constructor(
    readonly x: number,
    readonly y: number,
  ) {}

  static formalize(point: InformalPoint): Point {
    if (point instanceof Point) {
      return point;
    }
    return new Point(point.x, point.y);
  }

  move(length: number, angle: number) {
    return new Point(
      this.x + length * Math.cos(rad(angle)),
      this.y + length * Math.sin(rad(angle)),
    );
  }
}

function rad(deg: number) {
  return deg * (Math.PI / 180);
}
