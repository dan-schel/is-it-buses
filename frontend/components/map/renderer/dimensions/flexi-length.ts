import { nonNull, parseFloatNull } from "@dan-schel/js-utils";

export class FlexiLength {
  // Note: "min" == the compressed length, "max" == the stretched length. In
  // some rare cases, "max" may be less than "min"!
  constructor(
    readonly min: number,
    readonly max: number,
  ) {}

  amplify(amplification: number) {
    return this.min + (this.max - this.min) * amplification;
  }

  plus(other: FlexiLength): FlexiLength {
    return new FlexiLength(this.min + other.min, this.max + other.max);
  }

  minus(other: FlexiLength): FlexiLength {
    return new FlexiLength(this.min - other.min, this.max - other.max);
  }

  times(factor: number): FlexiLength {
    return new FlexiLength(this.min * factor, this.max * factor);
  }

  divide(factor: number): FlexiLength {
    return new FlexiLength(this.min / factor, this.max / factor);
  }

  toString() {
    const min = this.min.toFixed(2);
    const max = this.max.toFixed(2);
    return `${min} ${max}`;
  }

  static fromString(s: string) {
    const parsed = s
      .split(" ")
      .map((n) => parseFloatNull(n))
      .filter(nonNull);

    if (parsed.length !== 2) {
      return null;
    }

    return new FlexiLength(parsed[0], parsed[1]);
  }
}

export function flexi(min: number, max?: number) {
  return new FlexiLength(min, max ?? min);
}
