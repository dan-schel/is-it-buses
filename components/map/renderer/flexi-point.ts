import { FlexiLength } from "@/components/map/renderer/flexi-length";
import { Point } from "@/components/map/renderer/point";
import { nonNull, parseFloatNull } from "@dan-schel/js-utils";

export class FlexiPoint {
  constructor(
    readonly min: Point,
    readonly max: Point,
  ) {}

  amplify(amplification: number) {
    return {
      x: this.min.x + (this.max.x - this.min.x) * amplification,
      y: this.min.y + (this.max.y - this.min.y) * amplification,
    };
  }

  horizontalDistanceTo(other: FlexiPoint): FlexiLength {
    return new FlexiLength(
      Math.abs(this.min.x - other.min.x),
      Math.abs(this.max.x - other.max.x),
    );
  }

  verticalDistanceTo(other: FlexiPoint): FlexiLength {
    return new FlexiLength(
      Math.abs(this.min.y - other.min.y),
      Math.abs(this.max.y - other.max.y),
    );
  }

  pythagoreanDistanceTo(point: FlexiPoint): FlexiLength {
    const dx = this.horizontalDistanceTo(point);
    const dy = this.verticalDistanceTo(point);
    return new FlexiLength(
      Math.sqrt(dx.min * dx.min + dy.min * dy.min),
      Math.sqrt(dx.max * dx.max + dy.max * dy.max),
    );
  }

  plus(input: { x?: FlexiLength; y?: FlexiLength }): FlexiPoint {
    const x = input.x ?? new FlexiLength(0, 0);
    const y = input.y ?? new FlexiLength(0, 0);
    return new FlexiPoint(
      new Point(this.min.x + x.min, this.min.y + y.min),
      new Point(this.max.x + x.max, this.max.y + y.max),
    );
  }

  minus(input: { x?: FlexiLength; y?: FlexiLength }): FlexiPoint {
    const x = input.x ?? new FlexiLength(0, 0);
    const y = input.y ?? new FlexiLength(0, 0);
    return new FlexiPoint(
      new Point(this.min.x - x.min, this.min.y - y.min),
      new Point(this.max.x - x.max, this.max.y - y.max),
    );
  }

  move(length: FlexiLength, angle: number): FlexiPoint {
    return new FlexiPoint(
      this.min.move(length.min, angle),
      this.max.move(length.max, angle),
    );
  }

  toString() {
    const minX = this.min.x.toFixed(2);
    const minY = this.min.y.toFixed(2);
    const maxX = this.max.x.toFixed(2);
    const maxY = this.max.y.toFixed(2);
    return `${minX} ${minY} ${maxX} ${maxY}`;
  }

  static fromString(s: string) {
    const parsed = s
      .split(" ")
      .map((n) => parseFloatNull(n))
      .filter(nonNull);

    if (parsed.length !== 4) {
      return null;
    }

    return new FlexiPoint(
      new Point(parsed[0], parsed[1]),
      new Point(parsed[2], parsed[3]),
    );
  }
}

export function fp(min: [number, number], max?: [number, number]): FlexiPoint {
  const second = max ?? min;
  return new FlexiPoint(
    new Point(min[0], min[1]),
    new Point(second[0], second[1]),
  );
}
