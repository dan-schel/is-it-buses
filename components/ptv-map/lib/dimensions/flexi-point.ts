import { BakedPoint } from "../../baked-geometry";
import { FlexiLength, InformalFlexiLength } from "./flexi-length";
import { InformalPoint, Point } from "./point";

export type InformalFlexiPoint =
  | FlexiPoint
  | { min: InformalPoint; max: InformalPoint }
  | InformalPoint;

export class FlexiPoint {
  constructor(
    readonly min: Point,
    readonly max: Point,
  ) {}

  static formalize(point: InformalFlexiPoint): FlexiPoint {
    if (point instanceof FlexiPoint) {
      return point;
    } else if ("min" in point) {
      return new FlexiPoint(
        Point.formalize(point.min),
        Point.formalize(point.max),
      );
    } else {
      const _point = Point.formalize(point);
      return new FlexiPoint(_point, _point);
    }
  }

  horizontalDistanceTo(other: FlexiPoint): FlexiLength {
    return new FlexiLength(
      Math.abs(this.min.x - other.min.x),
      Math.abs(this.max.x - other.max.x),
    );
  }

  verticalDistanceTo(other: FlexiPoint): FlexiLength {
    return new FlexiLength(
      Math.abs(this.min.y - other.min.y),
      Math.abs(this.max.y - other.max.y),
    );
  }

  plus(input: {
    x?: InformalFlexiLength;
    y?: InformalFlexiLength;
  }): FlexiPoint {
    const x = FlexiLength.formalize(input.x ?? 0);
    const y = FlexiLength.formalize(input.y ?? 0);
    return new FlexiPoint(
      new Point(this.min.x + x.min, this.min.y + y.min),
      new Point(this.max.x + x.max, this.max.y + y.max),
    );
  }

  minus(input: {
    x?: InformalFlexiLength;
    y?: InformalFlexiLength;
  }): FlexiPoint {
    const x = FlexiLength.formalize(input.x ?? 0);
    const y = FlexiLength.formalize(input.y ?? 0);
    return new FlexiPoint(
      new Point(this.min.x - x.min, this.min.y - y.min),
      new Point(this.max.x - x.max, this.max.y - y.max),
    );
  }

  move(length: InformalFlexiLength, angle: number): FlexiPoint {
    const _length = FlexiLength.formalize(length);
    return new FlexiPoint(
      this.min.move(_length.min, angle),
      this.max.move(_length.max, angle),
    );
  }

  bake(): BakedPoint {
    return new BakedPoint(this.min.x, this.min.y, this.max.x, this.max.y);
  }
}
