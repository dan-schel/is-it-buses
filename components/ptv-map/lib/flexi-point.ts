import { FlexiLength, InformalFlexiLength } from "./flexi-length";
import { InformalPoint, Point } from "./point";

export type InformalFlexiPoint =
  | FlexiPoint
  | { min: InformalPoint; max: InformalPoint }
  | InformalPoint;

export class FlexiPoint {
  constructor(
    readonly min: Point,
    readonly max: Point,
  ) {}

  static formalize(point: InformalFlexiPoint): FlexiPoint {
    if (point instanceof FlexiPoint) {
      return point;
    } else if ("min" in point) {
      return new FlexiPoint(
        Point.formalize(point.min),
        Point.formalize(point.max),
      );
    } else {
      const _point = Point.formalize(point);
      return new FlexiPoint(_point, _point);
    }
  }

  horizontalDistanceTo(other: FlexiPoint): FlexiLength {
    return new FlexiLength(
      Math.abs(this.min.x - other.min.x),
      Math.abs(this.max.x - other.max.x),
    );
  }

  verticalDistanceTo(other: FlexiPoint): FlexiLength {
    return new FlexiLength(
      Math.abs(this.min.y - other.min.y),
      Math.abs(this.max.y - other.max.y),
    );
  }

  amplify(amplification: number): Point {
    return new Point(
      this.min.x + (this.max.x - this.min.x) * amplification,
      this.min.y + (this.max.y - this.min.y) * amplification,
    );
  }

  plus(input: {
    x?: InformalFlexiLength;
    y?: InformalFlexiLength;
  }): FlexiPoint {
    const x = FlexiLength.formalize(input.x ?? 0);
    const y = FlexiLength.formalize(input.y ?? 0);
    return new FlexiPoint(
      new Point(this.min.x + x.min, this.min.y + y.min),
      new Point(this.max.x + x.max, this.max.y + y.max),
    );
  }

  minus(input: {
    x?: InformalFlexiLength;
    y?: InformalFlexiLength;
  }): FlexiPoint {
    const x = FlexiLength.formalize(input.x ?? 0);
    const y = FlexiLength.formalize(input.y ?? 0);
    return new FlexiPoint(
      new Point(this.min.x - x.min, this.min.y - y.min),
      new Point(this.max.x - x.max, this.max.y - y.max),
    );
  }
}
