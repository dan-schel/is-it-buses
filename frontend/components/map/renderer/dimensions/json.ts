import { FlexiLength } from "@/frontend/components/map/renderer/dimensions/flexi-length";
import { FlexiPoint } from "@/frontend/components/map/renderer/dimensions/flexi-point";
import { nonNull } from "@dan-schel/js-utils";
import { z } from "zod";

export const flexiPointStringJson = schemaForCommaSeparated((x) =>
  FlexiPoint.fromString(x),
);

export const flexiLengthStringJson = schemaForCommaSeparated((x) =>
  FlexiLength.fromString(x),
);

export function createFlexiPointString(input: readonly FlexiPoint[]) {
  return toCommaSeparated(input);
}

export function createFlexiLengthString(input: readonly FlexiLength[]) {
  return toCommaSeparated(input);
}

function schemaForCommaSeparated<T extends object>(
  parser: (x: string) => T | null,
) {
  return z.string().transform((input, ctx) => {
    const numbers = input.split(",").map((x) => parser(x));
    const parsed = numbers.filter(nonNull);

    if (parsed.length !== numbers.length) {
      ctx.issues.push({
        code: "custom",
        message: "Not a valid comma separated string of numbers.",
        input,
      });
      return z.NEVER;
    }

    return parsed;
  });
}

function toCommaSeparated(input: readonly { toString: () => string }[]) {
  return input.map((x) => x.toString()).join(",");
}
