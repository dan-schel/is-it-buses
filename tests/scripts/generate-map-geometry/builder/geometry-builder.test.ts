import { describe, expect, it } from "vitest";
import { InterchangeBlueprint } from "@/scripts/generate-map-geometry/lib/blueprint/interchange-blueprint";
import { LineBlueprint } from "@/scripts/generate-map-geometry/lib/blueprint/line-blueprint";
import { fp } from "@/scripts/generate-map-geometry/lib/dimensions/flexi-point";
import { PathBlueprint } from "@/scripts/generate-map-geometry/lib/blueprint/path-blueprint";
import { flexi } from "@/scripts/generate-map-geometry/lib/dimensions/flexi-length";
import { GeometryBuilder } from "@/scripts/generate-map-geometry/lib/builder/geometry-builder";

describe("GeometryBuilder", () => {
  it("builds the geometry as expected", () => {
    const interchange = InterchangeBlueprint.simple(
      1,
      ["line1", "line2"],
      "line1",
      "left-edge",
      "line2",
      "right-edge",
    );

    const line1 = new LineBlueprint({
      origin: fp([0, 0]),
      angle: 0,
      color: "cyan",
      path: new PathBlueprint()
        .terminus()
        .straight(flexi(45, 90))
        .station(interchange.point("line1"))
        .straight(flexi(5))
        .split({
          split: new PathBlueprint()
            .curve(flexi(15), 45)
            .straight(flexi(25, 50))
            .terminus(),
        })
        .curve(flexi(10), -45)
        .straight(flexi(45, 90))
        .terminus(),
    });

    const line2 = new LineBlueprint({
      origin: fp([0, 5]),
      angle: 0,
      color: "purple",
      path: new PathBlueprint()
        .terminus()
        .straight(flexi(45, 90))
        .station(interchange.point("line2"))
        .straight(flexi(5))
        .curve(flexi(10), 45)
        .straight(flexi(45, 90))
        .terminus(),
    });

    const geometry = new GeometryBuilder().build([line1, line2]);

    expect(JSON.stringify(geometry.toJSON(), null, 2)).toMatchSnapshot();
  });
});
