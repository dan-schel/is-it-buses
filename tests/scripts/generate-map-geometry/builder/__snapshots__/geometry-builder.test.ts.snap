// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GeometryBuilder > builds the geometry as expected 1`] = `
"{
  "segments": [
    {
      "startNodeId": 1,
      "endNodeId": 2,
      "color": "cyan",
      "points": "0.00 0.00 0.00 0.00,45.00 0.00 90.00 0.00",
      "distances": "0.00 0.00,45.00 90.00"
    },
    {
      "startNodeId": 2,
      "endNodeId": 3,
      "color": "cyan",
      "points": "45.00 0.00 90.00 0.00,50.00 0.00 95.00 0.00,52.35 0.18 97.35 0.18,54.64 0.73 99.64 0.73,56.81 1.63 101.81 1.63,58.82 2.86 103.82 2.86,60.61 4.39 105.61 4.39,78.28 22.07 140.96 39.75",
      "distances": "0.00 0.00,5.00 5.00,7.35 7.35,9.71 9.71,12.06 12.06,14.42 14.42,16.77 16.77,41.77 66.77"
    },
    {
      "startNodeId": 2,
      "endNodeId": 4,
      "color": "cyan",
      "points": "45.00 0.00 90.00 0.00,50.00 0.00 95.00 0.00,51.56 -0.12 96.56 -0.12,53.09 -0.49 98.09 -0.49,54.54 -1.09 99.54 -1.09,55.88 -1.91 100.88 -1.91,57.07 -2.93 102.07 -2.93,88.89 -34.75 165.71 -66.57",
      "distances": "0.00 0.00,5.00 5.00,6.57 6.57,8.14 8.14,9.71 9.71,11.28 11.28,12.85 12.85,57.85 102.85"
    },
    {
      "startNodeId": 5,
      "endNodeId": 6,
      "color": "purple",
      "points": "0.00 5.00 0.00 5.00,45.00 5.00 90.00 5.00",
      "distances": "0.00 0.00,45.00 90.00"
    },
    {
      "startNodeId": 6,
      "endNodeId": 7,
      "color": "purple",
      "points": "45.00 5.00 90.00 5.00,50.00 5.00 95.00 5.00,51.56 5.12 96.56 5.12,53.09 5.49 98.09 5.49,54.54 6.09 99.54 6.09,55.88 6.91 100.88 6.91,57.07 7.93 102.07 7.93,88.89 39.75 165.71 71.57",
      "distances": "0.00 0.00,5.00 5.00,6.57 6.57,8.14 8.14,9.71 9.71,11.28 11.28,12.85 12.85,57.85 102.85"
    }
  ],
  "interchanges": [
    {
      "thick": [
        "45.00 -2.00 90.00 -2.00,45.00 7.00 90.00 7.00"
      ]
    }
  ],
  "termini": [
    {
      "color": "cyan",
      "points": "0.00 -5.00 0.00 -5.00,0.00 5.00 0.00 5.00"
    },
    {
      "color": "cyan",
      "points": "81.82 18.54 144.50 36.21,74.75 25.61 137.43 43.28"
    },
    {
      "color": "cyan",
      "points": "85.36 -38.28 162.18 -70.10,92.43 -31.21 169.25 -63.03"
    },
    {
      "color": "purple",
      "points": "0.00 0.00 0.00 0.00,0.00 10.00 0.00 10.00"
    },
    {
      "color": "purple",
      "points": "92.43 36.21 169.25 68.03,85.36 43.28 162.18 75.10"
    }
  ],
  "viewport": {
    "min": {
      "x": 44.44543648263006,
      "y": 2.4999999999999964,
      "w": 88.89087296526012,
      "h": 74.49747468305833
    },
    "max": {
      "x": 82.85533905932738,
      "y": 2.499999999999993,
      "w": 165.71067811865476,
      "h": 138.1370849898476
    }
  }
}"
`;
